### PR TITLE
aws_request_signing: Signature within query parameters 

### PR DESCRIPTION
**Pull request by nbaws**
Commit Message: aws_request_signing: Signature within query parameters 
Additional Description: Per design in https://github.com/envoyproxy/envoy/issues/32388
Risk Level: Low
Testing: Unit
Docs Changes: Docs updated
Release Notes: Updated
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] 32388
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

@suniltheta

### CODE REVIEW

**Review by abeyad — COMMENTED** [review]
Left comments on the API

/wait

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto [hunk 1]
```diff
@@ -32,6 +34,13 @@ message AwsRequestSigning {
     AWS_SIGV4A = 1;
   }
 
+  message QueryString {
+    // Optional expiration time for the query string parameters, in seconds. As query string parameter based requests are replayable, in effect representing
+    // an API call that has already been authenticated, it is recommended to keep this expiration time as short as feasible.
+    // This value will default to 5 seconds and has a maximum value of 3600 seconds (1 hour).
+    google.protobuf.UInt32Value expiration_time = 1;
```
[the remark this answers is not in the document]
Comment (nbaws) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:41]: updated. thanks for this!

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto [hunk 2]
```diff
@@ -87,6 +96,14 @@ message AwsRequestSigning {
 
   // Optional Signing algorithm specifier, either ``AWS_SIGV4`` or ``AWS_SIGV4A``, defaulting to ``AWS_SIGV4``.
   SigningAlgorithm signing_algorithm = 6;
+
+  // Use the query string to store output of SigV4 or SigV4A calculation, rather than HTTP headers. The ``Authorization`` header will not be modified if ``query_string``
+  // is configured
```
[the remark this answers is not in the document]
Comment (abeyad) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:101]: I'm fine with the wording but I do think it would be better to start the sentence with "If set, use the query string" ... 
That makes it clearer that this field is meant to trigger a certain behavior iff it is set.  

Also, if you can please end the sentence with a period (i.e. `.`).  

Thanks!

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto [hunk 3]
```diff
@@ -32,6 +34,13 @@ message AwsRequestSigning {
     AWS_SIGV4A = 1;
   }
 
+  message QueryString {
+    // Optional expiration time for the query string parameters. As query string parameter based requests are replayable, in effect representing
+    // an API call that has already been authenticated, it is recommended to keep this expiration time as short as feasible.
+    // This value will default to 5 seconds and has a maximum value of 3600 seconds (1 hour).
+    google.protobuf.Duration expiration_time = 1;
```
[the remark this answers is not in the document]
Comment (nbaws) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:41]: thank you for this. my protobuf knowledge is not strong!

**Review by abeyad — APPROVED** [review]
/lgtm api

thanks for addressing the feedback!

File: source/extensions/common/aws/signer_base_impl.cc [hunk 4]
```diff
@@ -111,6 +145,40 @@ std::string SignerBaseImpl::createContentHash(Http::RequestMessage& message, boo
...
+                     credential_scope);
+}
+
+void SignerBaseImpl::createQueryParams(Envoy::Http::Utility::QueryParamsMulti& query_params,
+                                       const absl::string_view credential,
+                                       const absl::string_view long_date,
+                                       const absl::optional<std::string> session_token,
+                                       const std::map<std::string, std::string>& signed_headers,
+                                       const uint8_t expiration_time) const {
+  // X-Amz-Algorithm
```
[the remark this answers is not in the document]
Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.cc:161]: now linked in signerbase along with the other AWS reference URL

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto [hunk 5]
```diff
@@ -32,6 +34,16 @@ message AwsRequestSigning {
     AWS_SIGV4A = 1;
   }
 
+  message QueryString {
+    // Optional expiration time for the query string parameters. As query string parameter based requests are replayable, in effect representing
+    // an API call that has already been authenticated, it is recommended to keep this expiration time as short as feasible.
+    // This value will default to 5 seconds and has a maximum value of 3600 seconds (1 hour).
+    google.protobuf.Duration expiration_time = 1 [(validate.rules).duration = {
```
[the remark this answers is not in the document]
Comment (nbaws) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:41]: In config.cc I now have

  uint16_t expiration_time =
      PROTOBUF_GET_SECONDS_OR_DEFAULT(config.query_string(), expiration_time, 5);

and a test case to validate this. It should allow user to use query_string: {} and it will default to 5 automatically

File: source/extensions/common/aws/signer_base_impl.cc [hunk 6]
```diff
@@ -94,10 +120,18 @@ void SignerBaseImpl::sign(Http::RequestHeaderMap& headers, const std::string& co
...
   // Phase 4: Sign request
-  const auto authorization_header = createAuthorizationHeader(
-      credentials.accessKeyId().value(), credential_scope, canonical_headers, signature);
-  ENVOY_LOG(debug, "Signing request with: {}", authorization_header);
-  headers.addCopy(Http::CustomHeaders::get().Authorization, authorization_header);
+  if (query_string_) {
+    // Append signature to existing query string
+    query_params.add(SignatureQueryParameters::get().AmzSignature, signature);
+    headers.setPath(query_params.replaceQueryString(headers.Path()->value()));
+    ENVOY_LOG(debug, "Query string appended to path");
```
[the remark this answers is not in the document]
Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.cc:127]: Good idea. Both are sanitised now.

File: test/extensions/filters/http/aws_request_signing/config_test.cc [hunk 7]
```diff
@@ -93,7 +93,92 @@ host_rewrite: new-host
...
+  )EOF";
+
+  AwsRequestSigningProtoConfig proto_config;
+  TestUtility::loadFromYamlAndValidate(yaml, proto_config);
+
+  AwsRequestSigningProtoConfig expected_config;
+  expected_config.set_service_name("s3");
+  expected_config.set_region("us-west-2");
+  expected_config.set_host_rewrite("new-host");
+  expected_config.mutable_query_string();
```
Comment (suniltheta) [inline test/extensions/filters/http/aws_request_signing/config_test.cc:116]: Do you think we can add this check?
`expected_config.mutable_query_string()->mutable_expiration_time()->set_seconds(5);`

File: source/extensions/common/aws/signer_base_impl.cc [hunk 8]
```diff
@@ -124,13 +125,24 @@ void SignerBaseImpl::sign(Http::RequestHeaderMap& headers, const std::string& co
...
+    auto sanitised_query_string = query_params.replaceQueryString(headers.Path()->value());
+    ENVOY_LOG(debug, "Query string signing - New path (sanitised): {}", sanitised_query_string);
+
   } else {
     const auto authorization_header = createAuthorizationHeader(
         credentials.accessKeyId().value(), credential_scope, canonical_headers, signature);
 
-    ENVOY_LOG(debug, "Signing request with: {}", authorization_header);
     headers.addCopy(Http::CustomHeaders::get().Authorization, authorization_header);
+    Envoy::Logger::Registry::setLogLevel(spdlog::level::debug);
```
Comment (suniltheta) [inline source/extensions/common/aws/signer_base_impl.cc:139]: why is this line required?

File: source/extensions/common/aws/signer_base_impl.cc [hunk 9]
```diff
@@ -124,13 +125,24 @@ void SignerBaseImpl::sign(Http::RequestHeaderMap& headers, const std::string& co
...
+
+    // Sanitize logged authorization header
+    std::vector<std::string> sanitised_header =
+        absl::StrSplit(authorization_header, absl::ByString("Signature="));
+    ENVOY_LOG(debug, "Header signing - Authorization header (sanitised): {}Signature=*****",
+              sanitised_header[0]);
```
Comment (suniltheta) [inline source/extensions/common/aws/signer_base_impl.cc:145]: if possible would you show an example of what this looks like before and after?

File: source/extensions/common/aws/signer_base_impl.cc [hunk 10]
```diff
@@ -124,13 +125,24 @@ void SignerBaseImpl::sign(Http::RequestHeaderMap& headers, const std::string& co
     // Append signature to existing query string
     query_params.add(SignatureQueryParameters::get().AmzSignature, signature);
     headers.setPath(query_params.replaceQueryString(headers.Path()->value()));
-    ENVOY_LOG(debug, "Query string appended to path");
+    // Sanitize logged query string
+    query_params.overwrite(SignatureQueryParameters::get().AmzSignature, "*****");
+    query_params.overwrite(SignatureQueryParameters::get().AmzSecurityToken, "*****");
```
Comment (suniltheta) [inline source/extensions/common/aws/signer_base_impl.cc:130]: Would it handle the condition if  `AmzSecurityToken` is not set on query_params?

Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.cc:130]: good catch ty

File: source/extensions/common/aws/signer_base_impl.cc (hunk 8 above)
Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.cc:139]: removed

File: source/extensions/common/aws/signer_base_impl.cc (hunk 9 above)
Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.cc:145]: After: 
Header signing - Authorization header (sanitised): AWS4-HMAC-SHA256 Credential=ASIAQ1234567/20240223/ap-southeast-2/vpc-lattice-svcs/aws4_request, SignedHeaders=accept;host;user-agent;x-amz-content-sha256;x-amz-date;x-amz-security-token;x-request-id, Signature=****

File: test/extensions/filters/http/aws_request_signing/config_test.cc (hunk 7 above)
Comment (nbaws) [reply test/extensions/filters/http/aws_request_signing/config_test.cc:116]: new test case added to verify that the default 5s expiration is being correctly set

File: changelogs/current.yaml [hunk 11]
```diff
@@ -212,6 +212,10 @@ new_features:
     added an option to dynamically set a per downstream connection idle timeout period object under the key
     ``envoy.tcp_proxy.per_connection_idle_timeout_ms``. If this filter state value exists, it will override the idle timeout
     specified in the filter configuration and the default idle timeout.
+- area: aws_request_signing
+  change: |
+    Update ``aws_request_signing`` filter to support optionally sending the aws signature in query parameters rather than headers,
+    by specifying the ``query_parameters`` configuration section.
```
Comment (suniltheta) [inline changelogs/current.yaml:218]: `s/query_parameters/query_string/`

Or preferably 
```
:ref:`query_string <envoy_v3_api_field_extensions.filters.http.aws_request_signing.v3.AwsRequestSigning.query_string>`
```

File: source/extensions/common/aws/signer_base_impl.cc [hunk 12]
```diff
@@ -111,6 +159,45 @@ std::string SignerBaseImpl::createContentHash(Http::RequestMessage& message, boo
...
+  // X-Amz-Expires
+  query_params.add(SignatureQueryParameters::get().AmzExpires, std::to_string(expiration_time));
+
+  // These three parameters can contain characters that require URL encoding
+  if (session_token.has_value()) {
+    // X-Amz-Security-Token
+
+    // TODO (@nbaws) : This should be using Utility::encodeQueryParam, but that function appears to
+    // be incorrectly double encoding = signs Will address this in a later patch and add test cases
+    // to ensure we haven't broken anything
```
Comment (suniltheta) [inline source/extensions/common/aws/signer_base_impl.cc:188]: I am not sure why format checker didn't flag for missing punctuation.

File: source/extensions/common/aws/signer_base_impl.h [hunk 13]
```diff
@@ -24,6 +24,22 @@ class SignatureHeaderValues {
...
+public:
+  // Query string parameters require camel case
+  const std::string AmzAlgorithm{"X-Amz-Algorithm"};
+  const std::string AmzCredential{"X-Amz-Credential"};
+  const std::string AmzDate{"X-Amz-Date"};
+  const std::string AmzRegionSet{"X-Amz-Region-Set"};
+  const std::string AmzSecurityToken{"X-Amz-Security-Token"};
+  const std::string AmzSignature{"X-Amz-Signature"};
+  const std::string AmzSignedHeaders{"X-Amz-SignedHeaders"};
+  const std::string AmzExpires{"X-Amz-Expires"};
```
Comment (suniltheta) [inline source/extensions/common/aws/signer_base_impl.h:37]: nit: is it possible to make these constexpr ?
similarly in SigV4ASignatureConstantValues as well

File: source/extensions/common/aws/signer_base_impl.h [hunk 14]
```diff
@@ -24,6 +24,22 @@ class SignatureHeaderValues {
...
+  const uint16_t DefaultExpiration = 5;
```
Comment (suniltheta) [inline source/extensions/common/aws/signer_base_impl.h:38]: nit: do you want to add code comments to indicate the unit as seconds?

**Review by suniltheta — COMMENTED** [review]
Thanks for making this change. Left few minor comments

File: source/extensions/common/aws/signer_base_impl.h (hunk 14 above)
Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.h:39]: addressed

File: source/extensions/common/aws/signer_base_impl.h (hunk 13 above)
Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.h]: modified all of these to constexpr

**Review by suniltheta — APPROVED** [review]

**Review by abeyad — DISMISSED** [review]
/lgtm api

File: source/extensions/common/aws/signer_base_impl.h [hunk 15]
```diff
@@ -24,6 +24,23 @@ class SignatureHeaderValues {
...
+  static constexpr char AmzAlgorithm[] = "X-Amz-Algorithm";
+  static constexpr char AmzCredential[] = "X-Amz-Credential";
+  static constexpr char AmzDate[] = "X-Amz-Date";
+  static constexpr char AmzRegionSet[] = "X-Amz-Region-Set";
+  static constexpr char AmzSecurityToken[] = "X-Amz-Security-Token";
+  static constexpr char AmzSignature[] = "X-Amz-Signature";
+  static constexpr char AmzSignedHeaders[] = "X-Amz-SignedHeaders";
+  static constexpr char AmzExpires[] = "X-Amz-Expires";
+  // Expiration time of query parameter request, in seconds
+  const uint16_t DefaultExpiration = 5;
```
Comment (wbpcode) [inline source/extensions/common/aws/signer_base_impl.h:39]: `constexpr abs::string_view` is prefered.

` static constexpr uint16_t DefaultExpiration = 5;`

File: source/extensions/common/aws/signer_base_impl.h [hunk 16]
```diff
@@ -24,6 +24,23 @@ class SignatureHeaderValues {
...
+};
+
+using SignatureQueryParameters = ConstSingleton<SignatureQueryParameterValues>;
```
Comment (wbpcode) [inline source/extensions/common/aws/signer_base_impl.h:42]: If `static constexpr` is used, then I think you needn't the `ConstSingleton`.

File: source/extensions/common/aws/sigv4_signer_impl.cc [hunk 17]
```diff
@@ -14,6 +14,7 @@
 #include "source/extensions/common/aws/utility.h"
 
 #include "absl/strings/str_join.h"
+#include "signer_base_impl.h"
```
Comment (wbpcode) [inline source/extensions/common/aws/sigv4_signer_impl.cc:17]: Please use the full path.

**Review by wbpcode — COMMENTED** [review]
Thanks for you contribution. I am not expert of aws signing, so I can only review the code rather than the logic. It's LGTM overall except some minor comments.

Will defer this to matt who are owner of this extension.

File: source/extensions/common/aws/signer_base_impl.h (hunk 16 above)
Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.h:42]: @wbpcode do you mean take these static constexpr out of the SignatureQueryParameterValues class entirely? ConstSingleton is how this class is instantiated.

Comment (wbpcode) [reply source/extensions/common/aws/signer_base_impl.h:42]: I mean if static `constexpr` is used, you may could access the value by `SignatureQueryParameterValues::xxxx` directly and needn't instance this class because all member are `static`?

Comment (nbaws) [reply source/extensions/common/aws/signer_base_impl.h:42]: ty - i did not know this. will clean these up!

**Comment by mattklein123** [comment]
LGTM can you merge main?

/wait

**Review by mattklein123 — APPROVED** [review]
